### ISSUE TITLE
Check the feasibility of using google forms and appscript instead of google tables form

### ISSUE THREAD

**Issue by aishwaryaharikumar** [hackforla/tables#53]
### Overview
We would like to use Google forms instead of Google Tables form in order to do data validation. 

### Details
The Google Table forms don't allow for any data validation and neither does the table itself.  Google Forms do allow for Regex validation, so if we could use the Google forms to capture the data, then we could keep people making errors in their submissions (e.g., "https://github.com/experimentsinhonesty" instead of the requested handle only "experimentsinhonesty"or "bonnieawolfe" instead of the requested linkedin url: "https://linkedin.com/in/bonnieawolfe")

### Action Items
- [x] make a simple google form
- [x] write an app script to get the data and populate it to the tables database
- [x] test to see if that works, and that the bots still fire for adding people to resources.
- [x] revise the form using the notes in the comments below from Harrison and Bonnie
- [x] Run by Product
- [x] Implement
- [ ] Run small beta test with 5 participants
- [ ] General rollout

#### Resources for working on this issue
- Use the test database, not the live one.
- API doc on sheet -> table: https://support.google.com/area120-tables/answer/10263946?hl=en&ref_topic=10081846&sjid=7922484318200369209-NC
- API doc on Google Table for appscript: https://developers.google.com/apps-script/advanced/tables

#### Resources created while working on this issue
- Google From [View](https://docs.google.com/forms/d/e/1FAIpQLSfCIkhCxf0q_wvt3XAh7P238YbCLSsKZ3zXzOo84RI_oT-63A/viewform) | [edit](https://docs.google.com/forms/d/1g_udE_AE1fW6o_v8cnyDZeK433bCIyMQbqDAE7F5Se4/edit) which is located in tables@hackforla.org google drive -> Scripts ->Onboarding
- [script](https://script.google.com/home/projects/1q005a-hE8YA1wOdRUIHSWawFVhSHkSm0a1MJ6CHeYBecSuSnMXeU8yhX/edit)
- [sheet](https://docs.google.com/spreadsheets/d/17vQpPrgdcakl-qVHDhKvChRheJyZMmFhyNDTaoPfeuY/edit?resourcekey#gid=1464356980)

**Comment by ExperimentsInHonesty** [comment]
@kennytram Please provide update
1. Progress: "What is the current status of your project? What have you completed and what is left to do?"
2. Blockers: "Difficulties or errors encountered."
3. Availability: "How much time will you have this week to work on this issue?"
4. ETA: "When do you expect this issue to be completed?"
5. Pictures or links* (if necessary): "Add any pictures or links that will help illustrate what you are working on."

* remember to add links to the top of the issue if they are going to be needed again.

**Comment by kennytram** [comment]
1. Progress: Added a validation within the script to correct github handle and linkedin automatically. 
2. Blockers: Been busy with other things lately.
3. Availability: I'll be busy for an unforeseeable amount of time
4. ETA: Someone else need to handle this issue

- Github regular expression to use to check if users are not using github.com/{username} (use Doesn't contain): 
github.com
- Linkedin regular exprssion to use to check if users are not entering the linkedin url properly (use Contains): 
linkedin.com/in/
- As far as I know, it doesn't seem possible to check whether the GitHub username (and LinkedIn) exists or not within Google Forms. Instead, we will need to submit the form and let the script handle the validation
- It's not possible to run the script while a user is filling out the form. 
- As for populating tables' database with Google Forms, that is definitely possible within Apps Script.

**Comment by wesley115** [comment]
1.Progress: "What is the current status of your project? What have you completed and what is left to do?"
2.Blockers:  
the only way to do real time validation for onboard form is add response validation for question

where for linkedin it would be ^https:\/\/www\.linkedin\.com\/in\/[a-zA-Z0-9_-]+\/?$ to guarantee valid urls

and for github it would be ^[a-zA-Z0-9_][a-zA-Z0-9_-]{0,38}[a-zA-Z0-9_]$ according to the github handle requirement.

the only way to validate whether the github handles and linkedin account exist or not is to do pull request from each API to check on script, however this won't be able to perform in real time, and can only be call once response sheet from google form is generated.


3.ETA: "When do you expect this issue to be completed?"

If checking handles and account existence with a script calling each API(not real time) is the approach, the script could be completed by next meeting

**Comment by wesley115** [comment]
https://tables.area120.google.com/
need internal validation for real time feedback from google tables database

**Comment by wesley115** [comment]
1.Progress:  

-Wrote an appscript that triggered by "google form" submit to sync the sheet "google form" generate to tables.

Reasons:
Since "google tables form" at the moment doesn't not have real time validation before submission, the only way to validate user's input before submission is through "google form"

- The script can get the data and populate it to the tables database, and the bots still fire for adding people to resources. It however will still need some refactoring once the format is confirmed

meeting update:
create formal google form
check feasibility with field with multiples labels
deploy during next meeting

**Comment by wesley115** [comment]
1.Progress: 
The script and form are ready for deployment
2.Blockers:
-Current google table API wont be able to sync edit from google form, instead it will create a new row. However with inserted regex for real time input validation in the google form, this should not be a big problem.
-appscript is not capable of creating lookup data type for a column which COP is currently using. Fix the label issue by changing it to Tag, which will still trigger the bot for adding to other table

The google sheet and google form is located at tables@hackforla.org google drive -> Scripts ->Onboarding

**Comment by wesley115** [comment]
The use of Google form can now replace Google Tables form for HFLA new member onboarding.

[script](https://script.google.com/home/projects/1q005a-hE8YA1wOdRUIHSWawFVhSHkSm0a1MJ6CHeYBecSuSnMXeU8yhX/edit)
[form](https://docs.google.com/forms/d/1g_udE_AE1fW6o_v8cnyDZeK433bCIyMQbqDAE7F5Se4/edit)
[sheet](https://docs.google.com/spreadsheets/d/17vQpPrgdcakl-qVHDhKvChRheJyZMmFhyNDTaoPfeuY/edit?resourcekey#gid=1464356980)

With every new submission from Google form "[HFLA Onboarding form](https://docs.google.com/forms/d/1g_udE_AE1fW6o_v8cnyDZeK433bCIyMQbqDAE7F5Se4/edit)", new data row will be created on the google sheet "HFLA Onboarding sheet" which are both located in
tables@hackforla.org's google drive -> Scripts -> Onboarding 

The script that import sheet to form is located in 
tables@hackforla.org's Apps Script-> import onboarding sheet to Google tables

Note that since the API doesn't support to sync the change between sheet and table, the script is now set to be triggered on form submission. Thus the Google form does not allow editing once submitted.

The "[Google form](https://docs.google.com/forms/d/1g_udE_AE1fW6o_v8cnyDZeK433bCIyMQbqDAE7F5Se4/edit)" is now performing data validation for GitHub Handle, LinkedIn and Passphrase.


_useful resources/reference_:

- API doc on sheet -> table: 
https://support.google.com/area120-tables/answer/10263946?hl=en&ref_topic=10081846&sjid=7922484318200369209-NC

- API doc on Google Table for appscript :
https://developers.google.com/apps-script/advanced/tables

**Comment by hkatzdev** [comment]
Seems good, just a few small things to fix.

a) The Linkedin field does not accept "https" urls without the www or any any "http" urls. It should also prevent "https://www.linkedin.com/in/name".
![Screenshot from 2024-01-10 15-13-30](https://github.com/hackforla/tables/assets/44281159/45126496-25a3-48e6-8a0d-5682c7488d62)
![Screenshot from 2024-01-10 15-16-50](https://github.com/hackforla/tables/assets/44281159/11d1326b-a8bf-4d15-960d-60f608b2720f)

b) The Slack verification message is a bit unclear to the user.
![Screenshot from 2024-01-10 15-11-02](https://github.com/hackforla/tables/assets/44281159/79a57df4-f6ad-484f-bad6-80e61745aff4)

c) Phone number is lacking validation.
![Screenshot from 2024-01-10 15-11-16](https://github.com/hackforla/tables/assets/44281159/fee401c8-3e48-43a6-9e1d-2f2228627d07)

d) Email is lacking validation. Ensure the validation is identical to Table's validation or it will error out.
![Screenshot from 2024-01-10 15-11-23](https://github.com/hackforla/tables/assets/44281159/df5099d7-2549-4105-b8ba-b53993cd9491)
![Screenshot from 2024-01-10 15-11-28](https://github.com/hackforla/tables/assets/44281159/dfca18fe-7252-4523-9923-5e0f52983f89)
![Screenshot from 2024-01-10 15-12-24](https://github.com/hackforla/tables/assets/44281159/35ba040c-b66b-45a8-a8c1-0861433a9be7)

**Comment by ExperimentsInHonesty** [comment]
## Tested the [form](https://docs.google.com/forms/d/e/1FAIpQLSfCIkhCxf0q_wvt3XAh7P238YbCLSsKZ3zXzOo84RI_oT-63A/viewform)

Here are my notes

### Phone # 
The form let me put in a phone number without enough digits.  There needs to  be a minimum and we should probably ask for a specific format.

### GitHub Handle
Maybe under the github handle we could put examples
❌ https://github.com/boniewolfe (this has more than just the handle)
❌ bonnie_wolfe (this has disallowed characters, GitHub handles only allow alphanumeric characters and dashes)
❌ bonnie wolfe (spaces are not allowed in GitHub usernames)
✔bonniewolfe 

Additionally, whatever GitHub handle we use, has to also be disallowed, to keep them from using the example instead of their handle.

### LinkedIn
disallow https://www.linkedin.com/in/name as a valid entry.

### Community of Practice
remove "HackforLA.org website"

### Time zone
This needs to be a required question
Let's change the wording to be: "What is your primary time zone?"

### After they submit
remove the ability for them to submit again 
Currently, it says

[Submit another response](https://docs.google.com/forms/u/0/d/e/1FAIpQLSfCIkhCxf0q_wvt3XAh7P238YbCLSsKZ3zXzOo84RI_oT-63A/viewform?usp=form_confirm)

**Comment by ExperimentsInHonesty** [comment]
@wesley115 I left you some notes on 2024-01-15.  Let us know if you are able to complete the changes requested.

**Comment by ExperimentsInHonesty** [comment]
@wesley115, we are removing you from this issue, so that we can put it back into the prioritized backlog.  If you want to come back to the team, let us know in the Slack.

**Comment by Reltre** [comment]
Progress: 
- Gained almost all context needed to work on the issue. 
- Updated form validation for LinkedIn URL to match requirements

Blockers: 
- I've been busy with other things this past week. But I plan to work on this some more tomorrow and on Friday.
Availability: I'm available for a couple of hours most weekdays to work on this.
- Some things I noted at least about the form validation. It seem somewhat limited. You can only add a single point of validation — containing or not containing text or a regex match — and lookaheads and lookbehinds are not supported, which is fairly limiting.

ETA: Remaining questions should be sorted by next week.

**Comment by ExperimentsInHonesty** [comment]
- [ ] check to see if it's possible to change the value for Accepting Responses to False (or whatever they are calling it).
 
![image](https://github.com/hackforla/tables/assets/37763229/161945ba-3be1-4dd7-8ef1-b4582082d051)

**Comment by ExperimentsInHonesty** [comment]
We probably need to make another issue for implementation.  Lets do that next week.

We will need the list of onboarding dates to be added to a table from the [2024 onboardings spreadsheet](https://docs.google.com/spreadsheets/d/1nfwWcwI8jCL_Ik4szvvlhCMUQsAkTDE1JdrG3xxFrB0/edit#gid=1681961916
)

add time
6:50 pm to 9:00 pm

cron in script runs on monday and tuesdays at 6:51 pm and confirms that there is an onboaring on that day, if yes, turns form on

**Comment by Reltre** [comment]
Progress:
* I finished updating the Form validations on the FE based on the two points of feedback above.
* I ran through the script as it is and tested it as well. It all seems to work as expected. There was one off by 1 error that was causing a field to not populate, but that should be fixed now.
* Next steps would be to update the script to populate the Tables workspace that is being used for signups. But, we should probably be careful doing this, maybe only try it out with a subset of people onboarding at first? 
* One other things that I noticed is that the Bound spreadsheet for the AppScript does actually populate and the data is instead sen to Tables. This isn't necessarily a bad thing, but it may be nice to have the Sheets data in case we want to use the Sheets API to migrate the data down the road.

Blockers:
* No blockers really. I think we should make a separate ticket/issue for adding in a function called within the `submit` hook to disable the form
* We also still need to consider adding in that column that would be used for disabling form submissions outside onboarding days

ETA: Effectively finished, but the script needs to be updated to point to the production Table. We can decide on how we want to do that going forward first.

**Comment by ExperimentsInHonesty** [comment]
Filling out this form will add you to Hack for LA resources.  This form should be used during onboarding only.

**Comment by Reltre** [comment]
- [ ] Add in a backend validation to the submit hook for the passphrase (validation should be `^[sS][tT][rR][aA][wW][bB][eE][rR][rR][yY]$`)

**Comment by ExperimentsInHonesty** [comment]
@Reltre during last night's test, it would not take anyone's phone number.  We tried
555-555-5555 and 5555555555 and 01-555-555-5555 and 015555555555
Please retest the form with expected inputs

**Comment by Reltre** [comment]
@bonniewolfe Weird, the regex page I was using was reporting a match, but I must have forgotten to run a manual test on that particular field. You're right in that it doesn't seem to work. My guess is that Google Forms doesn't support some type of regex that I'm using. I'll fix it up

**Comment by Reltre** [comment]
@ExperimentsInHonesty I believe I've fixed the issue. I tested the field with all test cases above and it was accepted in all instances.

**Comment by ExperimentsInHonesty** [comment]
@Reltre The form is working now, except for the community of practice.  I filled it out, and selected Product Management, but it was not recorded in the row on the tables (see row 2414)

**Comment by ExperimentsInHonesty** [comment]
@wesley115 and I were looking at [API doc on Google Table for appscript](https://developers.google.com/apps-script/advanced/tables) and this part seems relevant to the problem

Rich data
Linked data
Relationship | String | String: "tables/[LINKED_TABLE_ID]/rows/[LINKED_ROW_ID]"
Lookup | Depends on the source column type. | This field can’t be modified and will update with the linked value.
Summary | Depends on the source column type and summary function: List Values: Array | This field can’t be modified.



![image](https://github.com/user-attachments/assets/791d1824-5f86-4399-a3ef-18bc0598cc07)

**Comment by Reltre** [comment]
@ExperimentsInHonesty Good, so that's the last piece then. I think the issue is that we don't have the right row_ids in the app script. So, when we go to populate that table, the CoP cell doesn't work as expected.

As far as I can tell, there isn't a way to look up row ids in the UI for sheets or tables. But, I think one way we can sort this out is programmatically. We could use the Tables API to look up row ids and then store them in the submit function for later use.

**Comment by ExperimentsInHonesty** [comment]
@Reltre I think these are the row ids and the table where the values come from, so not sure what you are saying regarding not being able to look them up in the UI.  But either way, what is expected ETA for getting this last part of the form working?

![image](https://github.com/user-attachments/assets/3c5e5b75-b379-4f60-adbd-2342b5eeaeec)

**Comment by Reltre** [comment]
@ExperimentsInHonesty I think you're right. I updated the table id in the script and the row ids yesterday. Based on documentation for the Tables API, the information provided in our current AppScript should work for setting Lookup columns from Volunteers to CoP.

How did it go for the same sample of signups yesterday? Did their rows get populated as excepted? Or are there still issues? If it's the latter, then we should check out the logs for recent submissions to sort out what else could be causing issues with row creation from AppScript to Tables

**Comment by Reltre** [comment]
I've tested out a solution for this issue with the table workspace marked "Copy of Volunteers". It seems like we were not using the correct key to reference the lookup column, and that's why it wasn't being created. I ran a test and it seems to work as expected now.

<img width="1453" alt="test adding lookup column for volunteers table" src="https://github.com/user-attachments/assets/c65aa7f1-feb9-4584-949d-1bdda7e414f4">


I've updated the table ids, row ids, and relationship title in the App Script to refer to the production Volunteers table. But, yes values are still in the script and commented out for later usage.
@ExperimentsInHonesty Perhaps we can see if a small number of new sign-ups next week are able to use the new form.

**Comment by ExperimentsInHonesty** [comment]
@Reltre I just tested it by filling out the form.  Please look into the following

### LinkedIn
The regex for linkedin was supposed to exclude people using the example

https://www.linkedin.com/in/name
but allow ones like this https://www.linkedin.com/in/bonnieawolfe

![image](https://github.com/user-attachments/assets/142fef0a-eeb0-4a32-97d3-22b820aece6a)

If that is not possible, can you write back with what the limitation is.

### Not recording to the production volunteer table
I filled out the form, but it did not write to the production environment tables.  Please test it yourself by filling out the form and then logging into tables to look for the new record you just created.  Please do this after you confirm that its ready and before you tell me it's ready.

**Comment by Reltre** [comment]
@ExperimentsInHonesty Are you referring to this table as the production table? https://tables.area120.google.com/u/3/workspace/bHKoRlcCZyubeRsnie2OgH/table/a8Fi8xyvbbq0gWnJv5p_tB. Perhaps I had a link to the wrong table. But, I do see your test record there at the end, which corresponds to a recent execution log from yesterday.
If that isn't the production table then I'll have to update the URIs being used, and I've just made a mistake there

**Comment by Reltre** [comment]
Regarding the URL check for LinkedIn fields, I tried some different things out for a bit and I finally landed on a regex that I think will work. Google forms didn't make it easy because it doesn't allow positive or negative lookarounds. But, it seems to be working now.

<img width="477" alt="Screenshot 2024-08-29 at 2 47 58 PM" src="https://github.com/user-attachments/assets/c3952fe4-d9c3-4e65-b0a3-a564c8d03bf8">
<img width="429" alt="Screenshot 2024-08-29 at 2 47 39 PM" src="https://github.com/user-attachments/assets/c5cb0499-f814-4edd-856a-99f930dbeda1">

**Comment by ExperimentsInHonesty** [comment]
Hi @Reltre I see your responses 

### Linkedin
Im not sure about what you are saying re the field regex.  It was allowing for both of these

https://www.linkedin.com/in/name
https://www.linkedin.com/in/bonnieawolfe

I just tried it again, and it is still allowing this text, which we don't want it to.
https://www.linkedin.com/in/name

### Production Database
You do have the correct production Table DB URL.  It seemed to show up eventually (after the meeting), but not sure why it took so long.  Please look at the timestamp on when I filled out the form and compare it to the first attended showing up in tables for the record: 2024-08-27 15:27 and let us know what the total delay time was

### Community of Practice choice
I selected Product Management and DevOps and I only see Product in the Table DB.  Please check the record in the form to confirm that I did choose both and then figure out why it didnt record both in the table.

**Comment by Reltre** [comment]
@ExperimentsInHonesty 
I've looked into the issues you've listed above. Per the listed timestamps, it seems like everything happened in a timely manner per the timestamps, but if you experienced something different, then there could be some problem here that we need to look into 🤔. Just to make sure we're both using the same form, sheet, and table, I want to relink them in this comment.

[Form](https://docs.google.com/forms/d/e/1FAIpQLSfCIkhCxf0q_wvt3XAh7P238YbCLSsKZ3zXzOo84RI_oT-63A/viewform)
[Sheet](https://docs.google.com/spreadsheets/d/17vQpPrgdcakl-qVHDhKvChRheJyZMmFhyNDTaoPfeuY/edit?pli=1&gid=1464356980#gid=1464356980)
[Table](https://tables.area120.google.com/u/3/workspace/bHKoRlcCZyubeRsnie2OgH/table/a8Fi8xyvbbq0gWnJv5p_tB)

### LinkedIn

I double-checked this, and it still seems to be working to me. But, it's possible that what you experienced was the validation not occurring when you expected it to. The form validation seems to only trigger after the next "focus" or click and not after you've finished typing.

**_Validation doesn't trigger after I stop typing_**
<img width="587" alt="Screenshot 2024-09-04 at 7 48 04 PM" src="https://github.com/user-attachments/assets/d794735e-9614-4019-a51f-42b6685fc664">

**_Validation trigger on focus of checkbox on next question_**
<img width="733" alt="Screenshot 2024-09-04 at 7 48 50 PM" src="https://github.com/user-attachments/assets/3f88d974-669e-4adf-804a-bba6f4478301">

**_Successful pass of validation after entering a valid LinkedIn URL_**
<img width="671" alt="Screenshot 2024-09-04 at 7 49 26 PM" src="https://github.com/user-attachments/assets/ab2dafbe-2433-431f-8156-a4ba1f80535b">

### Production Database
Based on you comment above, it sounds like you filled out the form and tested it out during the last onboarding. I made sure to check the date of the last onboarding, and it was on the 20th of August. Now, interestingly enough, there was one record that was added a day later (unrelated to the test). And the record that you added to test this new Form was added on the 28th of August, 8 days later.

If you did fill out the form on the 20th between 6-8 PM, then an 8 day delay is really really strange. I checked the timestamps for your form submission and change history for the table and they do match — both occurring August 28th at 3:27 PM according to Google Forms and Tables.

**_Form Submission date and timestamp_**
<img width="812" alt="Screenshot 2024-09-04 at 7 43 49 PM" src="https://github.com/user-attachments/assets/3c938e09-b3b3-43dc-b1db-b625eed23aeb">

**_Tables change history timestamp_**
<img width="594" alt="Screenshot 2024-09-04 at 7 43 13 PM" src="https://github.com/user-attachments/assets/a758c16e-0f0c-4fb8-84d8-dc03ced6751b">

### Community of Practice Choice

I checked the most recent record and it does show both for me. Let me know if perhaps you see it now. There was a column that only shows "Product Manager" and that's the `job_title_target` column.

**_Most recent record added to production table showing both Product Manager and DevOps under CoP_**
<img width="953" alt="Screenshot 2024-09-04 at 7 44 42 PM" src="https://github.com/user-attachments/assets/300f3568-358d-43de-8ff4-358acfc597f5">

**Comment by ExperimentsInHonesty** [comment]
@Reltre I tried the form out on 1/2 the people last night, and the validation on the LinkedIn is not allowing dashes.  But dashes are in some LinkedIn profiles. Example: https://www.linkedin.com/in/allan-lopez-2ba1b0310.  Please fix and we will try again.

FYI: You don't need the following info, but in case you are wondering how I executed the A/B test, I created a 2 new slide decks.

- [Hack for LA Self Onboarding - A/B test](https://docs.google.com/presentation/d/1mZnGtvTpgn_H6YVmCh2377Sbvi9uDUFINT3cILE2FTc/) - this has you choose an A or B group based on your street address.  The slide decks it takes you to are identical except for one slide (the one with the form)
   - [A: Hack for LA Self Onboarding](https://docs.google.com/presentation/d/191qPSAGahj2af-4VBWBuYHH2tgLqrKHTjbDy9o6D7-0/edit#slide=id.g10f1395a4f0_0_97) - new form
   - [B: Hack for LA Self Onboarding](https://docs.google.com/presentation/d/1bueC40PkklLSJVPttiFeYJvcB69PgJx8pxZK5G7gzzo/edit?pli=1#slide=id.g10f1395a4f0_0_97) - old form

**Comment by Reltre** [comment]
@ExperimentsInHonesty Ok, I see, I didn't realize that hyphens were a possible option so I didn't include them in the regex. On the LinkedIn website, it says that a profile should only use letters and numbers, but maybe that's a recent change. I definitely see a lot of linkedin usernames with hyphens.

I've updated the regex to allow for hypens and I've tested the form as well for the possible inclusion of hyphens.
